### DISCUSSION TITLE
ERROR: onedrive application is already running - check system process list for active application instances

### DISCUSSION THREAD

**Discussion by rrodrigueznt** [abraunegg/onedrive#2070] [General]
Hi! First: thanks to abraunegg and all contributors for this great piece of software! Now, a question!

Running `onedrive v2.4.20` on `CentOS Stream 8` server. Catching up with all available features. I've several OneDrive services running. When I query a shared library name using the client to obtain the required configuration details, I get `ERROR. onedrive application is already running`. I have to use a different use to execute the command. For instance, `sudo /usr/local/bin/onedrive --get-O365-drive-id 'IGFAE-SC'` do the trick.

Please, why I can run several OneDrive instances and I can't execute 'onedrive' to get the properties of a shared library?

Thanks for your time!

**Comment by abraunegg** [comment]
@rrodrigueznt 
> Hi! First: thanks to abraunegg and all contributors for this great piece of software!

Thanks for the feedback - greatly apprecaited.

> Now, a question! ... Please, why I can run several OneDrive instances and I can't execute 'onedrive' to get the properties of a shared library?

The easiest way to explain what is going on - is to explain how the application works.

By default, the application will use `~/.config/onedrive/` to store all it's runtime data in - that is configuration files, hashes of those files, and, more importantly, the local cache database file that details what files are being accessed, their sync state and so on. The client as specific protections built in, so that only 1 runtime can use the same configuration path - otherwise, what happens is - the cache file gets all sorts of bad data from the multiple configurations - and - can lead to a very messy end-state including data corruption | data loss ... this is done for your protection. This is why you get `ERROR: onedrive application is already running - check system process list for active application instances` - as you have 2 application processes trying to access the same configuration data.

So with this being said, when you configure additional services, you pass in the `--confdir` parameter - hopefully using a *different* path than the default - so hopefully in your case - something similar to this:

* `onedrive --confdir="~/.config/shared_library_1" <any other options>`
* `onedrive --confdir="~/.config/shared_library_2" <any other options>`
* `onedrive --confdir="~/.config/shared_library_3" <any other options>`
...
* `onedrive --confdir="~/.config/shared_library_99" <any other options>`

What this means is - all your Shared Libraries are using their own configuration path, rather than the 'default' path.

Now - why cant you run the required command .. essentially, one of your existing configurations is using the 'default' path ... you need to re-evaluate all your configurations and work out which one is using that default path, change it, and then you will be able to just run the client to get the appropriate path you need.

Now ... I also note that you are running the client with `sudo` ... please do not do that.

What you are doing is running the client with 'root' privileges - rather than running this using your *normal* user account.

When you use sudo - you are also using the 'root' home directory for all your configuration data, configuration files, and - when all your data is being downloaded from OneDrive, it is being downloaded with 'root' only permissions ... meaning then you are having to do some permission mangling to maybe even access the data as an unprivileged user ...

Please stop running the client via sudo

**Reply by rrodrigueznt** [reply]
Thanks! I greatly appreciate your warning about not running the client with `sudo`. I understand that it is a dangerous alternative. But, in spite of. I _only_ run it to get information about a library, not to synchronization actions.

When I have other running instances/services, I can run `onedrive --version`, but I can not run `onedrive --get-O365-drive-id 'my-shared-library'`.

Does `--get-O365-drive-id 'my-shared-library'` use in any way the configuration files?

**Comment by abraunegg** [comment]
@rrodrigueznt 
>  Thanks! I greatly appreciate your warning about not running the client with sudo. I understand that it is a dangerous alternative. But, in spite of. I only run it to get information about a library, not to synchronization actions.

Hrmm .. well something is not right then - as, if you are getting that error message - you *must* have an active running configuration running under your 'root' account somewhere ......

> Does --get-O365-drive-id 'my-shared-library' use in any way the configuration files?

Yes it does - because it has to authenticate with the OneDrive API to perform the applicable queries to get the correct drive id for the library you are querying.

**Reply by rrodrigueznt** [reply]
@abraunegg 
> @rrodrigueznt
> 
> > Thanks! I greatly appreciate your warning about not running the client with sudo. I understand that it is a dangerous alternative. But, in spite of. I only run it to get information about a library, not to synchronization actions.
> 
> Hrmm .. well something is not right then - as, if you are getting that error message - you _must_ have an active running configuration running under your 'root' account somewhere ......

I'm not able to find it. All, currently four, services run at startup, and all with my username as parameter in the form onedrive@username.service
> 
> > Does --get-O365-drive-id 'my-shared-library' use in any way the configuration files?
> 
> Yes it does - because it has to authenticate with the OneDrive API to perform the applicable queries to get the correct drive id for the library you are querying.

But when I use '--get-O365-drive-id` switch I always get asked to authenticate with the OneDrive API, so it seems that it doesn't store the login information in anyway.

**Reply by abraunegg** [reply]
@rrodrigueznt 
> I'm not able to find it. All, currently four, services run at startup, and all with my username as parameter in the form

This is where 'sudo' *does* come in handy:
```
sudo ps aufxw | grep onedrive
```

This will show all running instances of 'onedrive' - and they should match what you think you have running.

> But when I use '--get-O365-drive-id` switch I always get asked to authenticate with the OneDrive API, so it seems that it doesn't store the login information in anyway

If you keep using sudo - you keep trying to authenticate with a 'root' instance ... dont. Please stop doing this.

Please try this:
1. Authenticate the client using your normal user (not root, no sudo): `onedrive --reauth`
2. Validate that in your `~/.config/onedrive/` folder, there is now configuration data / information
3. Query for your Shared Library: `onedrive --get-O365-drive-id 'my-shared-library'`

**Reply by rrodrigueznt** [reply]
> @rrodrigueznt
> 
> > I'm not able to find it. All, currently four, services run at startup, and all with my username as parameter in the form
> 
> This is where 'sudo' _does_ come in handy:
> 
> ```
> sudo ps aufxw | grep onedrive
> ```

Here the whole output:

```
(base) [rjr@e220i059h057w ~]$ sudo ps aufxw | grep onedrive
rjr       122555  0.0  0.0 221944  1188 pts/2    S+   14:28   0:00              \_ grep --color=auto onedrive
rjr       114546  0.4  0.1 656792 39448 ?        Ssl  Jul30  21:21 /usr/local/bin/onedrive --monitor --confdir=/home/rjr/.config/onedrive_personal
rjr       120003  4.3  0.1 687560 64092 ?        Ssl  Aug01  41:16 /usr/local/bin/onedrive --monitor --confdir=/home/rjr/.config/onedrive
rjr       120689  4.4  0.1 687580 64156 ?        Ssl  08:27  16:12 /usr/local/bin/onedrive --monitor --confdir=/home/rjr/.config/onedrive_usc
rjr       121784  0.0  0.0 637004 15028 ?        Ssl  12:21   0:00 /usr/local/bin/onedrive --monitor --confdir=/home/rjr/.config/onedrive_igfae-sc
(base) [rjr@e220i059h057w ~]$
```
It perfectly matched what I expected. No instance is running under `root`; all of them run under my user, `rjr` 

> 
> This will show all running instances of 'onedrive' - and they should match what you think you have running.
> 
> > But when I use '--get-O365-drive-id` switch I always get asked to authenticate with the OneDrive API, so it seems that it doesn't store the login information in anyway
> 
> If you keep using sudo - you keep trying to authenticate with a 'root' instance ... dont. Please stop doing this.
> 
> Please try this:
> 
> 1. Authenticate the client using your normal user (not root, no sudo): `onedrive --reauth`
> 2. Validate that in your `~/.config/onedrive/` folder, there is now configuration data / information
> 3. Query for your Shared Library: `onedrive --get-O365-drive-id 'my-shared-library'`

It worked as expected althought between points 1 and 2 I've to cancel the execution of the command as it asked for re-authorization:

```
(base) [rjr@e220i059h057w onedrive]$ onedrive --reauth
Configuration file successfully loaded
Deleting the saved authentication status ... re-authentication requested
Configuring Global Azure AD Endpoints
Authorize this app visiting: ...
```
After 3, authorization information is stored again in the concerned config folder.

But it is not what I'm trying to get. All what I want is a way of running the `onedrive` client to read the `drive_id` of a shared library to create an ad hoc configuration and service. The method described in the [documentation](https://github.com/abraunegg/onedrive/blob/master/docs/BusinessSharedFolders.md
) works great if the default configuration is not in use yet. If anybody, like me, is using the default configuration will face the error in the topic of this discussion.

Thanks!

**Comment by abraunegg** [comment]
@rrodrigueznt 
This is using the 'default' configuration folder:
```
rjr       120003  4.3  0.1 687560 64092 ?        Ssl  Aug01  41:16 /usr/local/bin/onedrive --monitor --confdir=/home/rjr/.config/onedrive
```

Update | Modify this configuration / setup to use a different configuration folder - then, you will not run into the issue.

~~This is your only solution - as - the client, to get the `drive_id` - the API has to be initialised ... it cant do that if the configuration is already in use.~~

**Reply by rrodrigueznt** [reply]
I get the explanation. Thanks. Thus, should we add some note at the beginning of [this section](https://github.com/abraunegg/onedrive/blob/master/docs/SharePoint-Shared-Libraries.md#query-that-shared-library-name-using-the-client-to-obtain-the-required-configuration-details) warning users about the need of creating an _ad-hoc_ configuration file to get configuration details in case the default one is in use?

**Comment by abraunegg** [comment]
@rrodrigueznt 
Whilst I work on a 'work-around' for this ... in this particular scenario, you can also add the `--dry-run`  parameter to your command.

This will allow you to query the required data if you are getting the 'onedrive application is already running' error message.

**Comment by abraunegg** [comment]
@rrodrigueznt 
Back to the original question:

> Please, why I can run several OneDrive instances and I can't execute 'onedrive' to get the properties of a shared library?

This is because the `--get-O365-drive-id` option requires the application database & sync engine to be initialised in order to query the OneDrive API for the SharePoint Site being queried.

PR https://github.com/abraunegg/onedrive/pull/2076 changes how `--get-O365-drive-id` uses the database & sync engine, so that you should now be able to use `--get-O365-drive-id` against a running instance, avoiding the error message saying that the client is already running or in use.

If you rebuild your client from 'master' you will get this fix today. This fix will be included in the v2.4.21 release.

**Reply by rrodrigueznt** [reply]
It works like a charm. I get the point of the solution. Thanks.
